[analyzer] Remove alpha.core.IdenticalExpr Checker

vabridgers opened the pull request:
This change removes the alpha.core.IdenticalExpr static analysis checker since it's checks are present in the clang-tidy checks misc-redundant-expression and bugprone-branch-clone. This check was implemented as a static analysis check using AST matching, and since alpha and duplicated in 2 clang-tidy checks may be removed.

NagyDonat reviewed (commented):
Thanks for implementing this cleanup commit!

I like that you ported the tests from the old checker to the tidy checks. However, as these test files are very large, perhaps it would be better to put these moved tests into stand-alone files instead of adding them at the end of existing files. (@ tidy developers: what would you prefer?)

Do I understand it correctly that the two tidy checkers report all the issues that were reported by alpha.core.IdenticalExpr? (Unfortunately github cannot show a clear diff between the old and new version of the moved code.)

carlosgalvezp commented:
> (@ tidy developers: what would you prefer?)

I agree!

vabridgers commented:
Hi @NagyDonat, I can port to new and smaller tests - no problem. This was an initial, speculative PR to get the conversation going, especially for comments like this. I'll update the PR. 

I did take the cases one by one from identical-expressions.cpp to the tidy checks and noticed one pattern. It looks to me like identical expressions utilizing floats were excluded in the tidy checks but found in the static analysis check. Do you want to retain that through improvements in misc-redundant-expression, or leave that marked as FIXME? I'm happy to take a look at an improvement like that. 

Thanks! - Vince

NagyDonat commented:
> I did take the cases one by one from identical-expressions.cpp to the tidy checks and noticed one pattern. It looks to me like identical expressions utilizing floats were excluded in the tidy checks but found in the static analysis check. Do you want to retain that through improvements in misc-redundant-expression, or leave that marked as FIXME? I'm happy to take a look at an improvement like that.

If it's not too difficult, then please improve the tidy checks to cover everything that was covered by the static analyzer checker that we want to remove. (Obviously this doesn't have a major practical importance as I assume that practically nobody used this alpha checker, but it's still nice to say that we preserve every feature.)

_(Disclaimer: if you think that the analyzer-only features happen to be buggy and/or completely useless, then don't port them.)_

If the changes are complex and you want to create several separate commits, then I think the natural commit order would be (1) adding the new features in tidy and then (2) removing the analyzer checker in a follow-up commit. However, I'd guess that it's reasonable to do these all in a single commit.

vabridgers commented:
@NagyDonat , sounds good. Thank you!

vabridgers committed:
[analyzer] Port alpha.core.IdenticalExpr to Tidy checks and remove

This change removes the alpha.core.IdenticalExpr static analysis
checker since it's checks are present in the clang-tidy checks
misc-redundant-expression and bugprone-branch-clone. This check was
implemented as a static analysis check using AST matching, and since
alpha and duplicated in 2 clang-tidy checks may be removed. The
existing LIT test was checked case by case, and the tidy checks
were improved to maintain alpha.core.IdenticalExpr features.

vabridgers commented:
I've reworked the test port in a "scaffolded" way so reviewers may see how the cases were considered one by one, and the existing tidy checkers improved to address cases IdenticalExpr was detecting and the tidy checkers were not. I needed to weaken the misc-redundant-expression AST matcher a bit, and consider an extra case for the bugprone-branch-clone checker. This picked up the integer and floating point literal cases, and also the duplicate outer/inner if cases. 

I'm happy to take next suggested steps to drive this improvement to completion, and would like some guidance on if the additional "scaffolded" case should be kept in place (to maintain clear differences from the original to new file in the git diff), or split the file now or maybe in a future separate commit so the differences are clear. 

Thanks - Vince

EugeneZelenko commented on clang-tools-extra/clang-tidy/bugprone/BranchCloneCheck.cpp:
```diff
@@ -102,6 +102,209 @@ void BranchCloneCheck::registerMatchers(MatchFinder *Finder) {
+  if (Stmt1->getStmtClass() != Stmt2->getStmtClass())
+    return false;
+
+  const Expr *Expr1 = dyn_cast<Expr>(Stmt1);
```
```suggestion
  const auto *Expr1 = dyn_cast<Expr>(Stmt1);
```

Type is explicitly stated in same statement.

EugeneZelenko commented on clang-tools-extra/clang-tidy/bugprone/BranchCloneCheck.cpp:
```diff
@@ -102,6 +102,209 @@ void BranchCloneCheck::registerMatchers(MatchFinder *Finder) {
+    return false;
+
+  const Expr *Expr1 = dyn_cast<Expr>(Stmt1);
+  const Expr *Expr2 = dyn_cast<Expr>(Stmt2);
```
```suggestion
  const auto *Expr2 = dyn_cast<Expr>(Stmt2);
```

EugeneZelenko commented on clang-tools-extra/clang-tidy/bugprone/BranchCloneCheck.cpp:
```diff
@@ -102,6 +102,209 @@ void BranchCloneCheck::registerMatchers(MatchFinder *Finder) {
+    return CastExpr1->getTypeAsWritten() == CastExpr2->getTypeAsWritten();
+  }
+  case Stmt::ReturnStmtClass: {
+    const ReturnStmt *ReturnStmt1 = cast<ReturnStmt>(Stmt1);
```
```suggestion
    const auto *ReturnStmt1 = cast<ReturnStmt>(Stmt1);
```

EugeneZelenko commented on clang-tools-extra/clang-tidy/bugprone/BranchCloneCheck.cpp:
```diff
@@ -102,6 +102,209 @@ void BranchCloneCheck::registerMatchers(MatchFinder *Finder) {
+  }
+  case Stmt::ReturnStmtClass: {
+    const ReturnStmt *ReturnStmt1 = cast<ReturnStmt>(Stmt1);
+    const ReturnStmt *ReturnStmt2 = cast<ReturnStmt>(Stmt2);
```
```suggestion
    const auto *ReturnStmt2 = cast<ReturnStmt>(Stmt2);
```

Same for other similar statements.

EugeneZelenko commented on clang-tools-extra/clang-tidy/misc/RedundantExpressionCheck.cpp:
```diff
@@ -1238,6 +1235,59 @@ void RedundantExpressionCheck::check(const MatchFinder::MatchResult &Result) {
+    const Expr *RHS = BinOp->getRHS()->IgnoreParenImpCasts();
+    BinaryOperator::Opcode Op = BinOp->getOpcode();
+
+    const DeclRefExpr *DeclRef1 = dyn_cast<DeclRefExpr>(LHS);
```
```suggestion
    const auto *DeclRef1 = dyn_cast<DeclRefExpr>(LHS);
```

Same below.

EugeneZelenko commented on clang-tools-extra/clang-tidy/bugprone/BranchCloneCheck.cpp:
```diff
@@ -102,6 +102,209 @@ void BranchCloneCheck::registerMatchers(MatchFinder *Finder) {
+  case Stmt::NullStmtClass:
+    return true;
+  case Stmt::CStyleCastExprClass: {
+    const CStyleCastExpr *CastExpr1 = cast<CStyleCastExpr>(Stmt1);
```
```suggestion
    const auto *CastExpr1 = cast<CStyleCastExpr>(Stmt1);
```

Same in many other similar statements below.

HerrCai0907 reviewed (commented):
looks good. it extends the checker in clang-tidy so i think documents and releasenote in clang-tidy are also needed

vabridgers commented:
Thanks @HerrCai0907, I'll update the documents and release notes. Best

5chmidti reviewed (commented):
> some guidance on if the additional "scaffolded" case should be kept in place (to maintain clear differences from the original to new file in the git diff), or split the file now or maybe in a future separate commit so the differences are clear.

IMO it is better as a follow-up so that the diff of the commit stays readable, which can also remove duplicate tests.

vabridgers commented on clang-tools-extra/test/clang-tidy/checkers/bugprone/alpha-core-identicalexpr.cpp:
I'll follow up after this PR is merged with a subsequent PR to split this LIT. If that's not acceptable please reopen this comment. Thanks.

vabridgers commented:
Thanks @5chmidti for the constructive comments. I believe those have been addressed. I'll follow up after this PR is approved and merged to split the LIT appropriately. Thanks for understanding this approach, it was the only one I could think of where the reviewers could transparently observe all features were maintained from the older checker. I'll happily make any other suggested improvements if I've missed something. Best

HerrCai0907 commented on clang-tools-extra/clang-tidy/bugprone/BranchCloneCheck.cpp:
```diff
@@ -102,6 +102,210 @@ void BranchCloneCheck::registerMatchers(MatchFinder *Finder) {
       this);
   Finder->addMatcher(switchStmt().bind("switch"), this);
   Finder->addMatcher(conditionalOperator().bind("condOp"), this);
+  Finder->addMatcher(ifStmt(hasDescendant(ifStmt())).bind("ifWithDescendantIf"),
```
`hasDescendant` will match lots of unrelated cases which will slow the speed. move the later check to here

vabridgers replied:
Hello @HerrCai0907 , I do not fully understand your suggestion. Could you describe what you mean by "move the later check to here"? Thanks for taking the time to review. Best!

5chmidti replied:
The current matcher will check for `if`s as descendants, not only as direct children. So code like
```c++
void foo() {
    if (true) {
        []() {
            if (false) {
            }
        };
    }
}
```

will still produce a match, that you filter out later, instead of in the matcher.
A matcher like `ifStmt(anyOf(has(ifStmt().bind("inner_if")), has(compoundStmt(has(ifStmt().bind("inner_if"))))).bind("ifWithDescendantIf")` will ensure that only direct children are looked at. To enforce the requirement of the first statement inside a compound statement, adding a custom matcher would be best IMO.

vabridgers replied:
Thanks @5chmidti. I think I had resolved the comment from @HerrCai0907 by narrowing the match. Would you agree that's sufficient, or would you like to see an improvement in this area before moving forward here? Thanks

HerrCai0907 commented on clang-tools-extra/clang-tidy/bugprone/BranchCloneCheck.cpp:
```diff
@@ -269,6 +473,21 @@ void BranchCloneCheck::check(const MatchFinder::MatchResult &Result) {
   }
 
+  if (const auto *IS = Result.Nodes.getNodeAs<IfStmt>("ifWithDescendantIf")) {
+    const Stmt *Then = IS->getThen();
+    auto CS = dyn_cast<CompoundStmt>(Then);
+    if (CS && (!CS->body_empty())) {
+      const auto *InnerIf = dyn_cast<IfStmt>(*CS->body_begin());
+      if (InnerIf && isIdenticalStmt(Context, IS->getCond(), InnerIf->getCond(),
+                                     /*IgnoreSideEffects=*/false)) {
```
I mean we can do this check in ast_matcher to reduce the cost

vabridgers commented:
Thanks @HerrCai0907 and @5chmidti for the constructive comments. I believe all have been addressed and a way forward has been described. Please let me know how we can move this forward. Thank you!

vabridgers commented:
Ping! Any more comments on this change? Thanks

HerrCai0907 commented on clang-tools-extra/clang-tidy/bugprone/BranchCloneCheck.cpp:
```diff
@@ -102,6 +102,211 @@ void BranchCloneCheck::registerMatchers(MatchFinder *Finder) {
+  if (Expr1 && Expr2) {
+    // If Stmt1 has side effects then don't warn even if expressions
+    // are identical.
+    if (!IgnoreSideEffects && Expr1->HasSideEffects(Ctx))
```
How about Stmt2 has side effects?

HerrCai0907 commented on clang-tools-extra/clang-tidy/bugprone/BranchCloneCheck.cpp:
```diff
@@ -102,6 +102,211 @@ void BranchCloneCheck::registerMatchers(MatchFinder *Finder) {
+    Expr::const_child_iterator I1 = Expr1->child_begin();
+    Expr::const_child_iterator I2 = Expr2->child_begin();
+    while (I1 != Expr1->child_end() && I2 != Expr2->child_end()) {
+      if (!*I1 || !*I2 || !isIdenticalStmt(Ctx, *I1, *I2, IgnoreSideEffects))
```
`!*I1 || !*I2` already be handled in `isIdenticalStmt`

vabridgers commented:
Hi @HerrCai0907 and @5chmidti , I see that @HerrCai0907 approved the review. Is it ok if I merge the change? Just wanted to make sure I'm not missing any comments specifically from @5chmidti . Thanks!
